File: Medium/51InvertBinaryTree.js
// O(n) time | O(n) space
// function invertBinaryTree(tree) {
//   // Write your code here.
//   let queue = [tree];
//   while (queue.length > 0) {
//     let current = queue.shift();
//     if (current === null) continue;

//     swapLeftAndRight(current);
//     queue.push(current.left);
//     queue.push(current.right);
//   }
// }

// O(n) time | O(d) space
function invertBinaryTree(tree) {
  if (tree === null) return;

  swapLeftAndRight(tree);
  invertBinaryTree(tree.left);
  invertBinaryTree(tree.right);
}
function swapLeftAndRight(tree) {
  let temp = tree.left;
  tree.left = tree.right;
  tree.right = temp;
}

// This is the class of the input binary tree.
class BinaryTree {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

File: Medium/51InvertBinaryTree.py
# # O(n) time | O(n)space
# def invertBinaryTree(tree):
#     # Write your code here.
#     queue = [tree]
#     while len(queue):
#         current = queue.pop(0)
#         if current is None:
#             continue
#         swapLeftAndRight(current)
#         queue.append(current.left)
#         queue.append(current.right)

# # O(n) time | O(d)space
def invertBinaryTree(tree):
    if tree == None:
        return

    swapLeftAndRight(tree)
    invertBinaryTree(tree.left)
    invertBinaryTree(tree.right)


def swapLeftAndRight(tree):
    tree.left, tree.right = tree.right, tree.left


# This is the class of the input binary tree.
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
